controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,283 @@
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	httpHeaders, err := r.getHTTPHeaders(*subscription)
+
+	if err != nil {
+		reqLogger.Error(err, "failed to get http headers")
+		return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+	}
+
+	prov, ready, err := r.getProvisioner(request, host)
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if err := r.createSubscription(ctx, prov, subscription, httpHeaders); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to create or update subscription")
+		}
+	} else {
+		// Being deleted
+		if err := r.deleteSubscription(ctx, prov, subscription); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to delete subscription")
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	return
+
+}
+
+func (r *BMCEventSubscriptionReconciler) createSubscription(ctx context.Context, prov provisioner.Provisioner, subscription *metal3v1alpha1.BMCEventSubscription, headers []map[string]string) error {
+	reqLogger := r.Log.WithName("bmceventsubscription")
+
+	if subscription.Status.SubscriptionID != "" {
+		reqLogger.Info("subscription already exists")
+		return nil
+	}
+
+	if _, err := prov.AddBMCEventSubscriptionForNode(subscription, headers); err != nil {
+		return errors.Wrap(err, "failed to create subscription")
+	}
+
+	err := r.Status().Update(ctx, subscription)
+
+	return err
+}
+
+func (r *BMCEventSubscriptionReconciler) deleteSubscription(ctx context.Context, prov provisioner.Provisioner, subscription *metal3v1alpha1.BMCEventSubscription) error {
+	reqLogger := r.Log.WithName("bmceventsubscription")
+	reqLogger.Info("deleting subscription")
+
+	if subscriptionHasFinalizer(subscription) {
+		if _, err := prov.RemoveBMCEventSubscriptionForNode(*subscription); err != nil {
+			return errors.Wrap(err, "failed to remove a subscription")
+		}
+
+		// Remove finalizer to allow deletion
+		subscription.Finalizers = utils.FilterStringFromList(
+			subscription.Finalizers, metal3v1alpha1.BMCEventSubscriptionFinalizer)
+		reqLogger.Info("cleanup is complete, removed finalizer",
+			"remaining", subscription.Finalizers)
+		if err := r.Update(context.Background(), subscription); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *BMCEventSubscriptionReconciler) getProvisioner(request ctrl.Request, host *metal3v1alpha1.BareMetalHost) (prov provisioner.Provisioner, ready bool, err error) {
+	reqLogger := r.Log.WithValues("bmceventsubscription", request.NamespacedName)
+
+	bmcCreds, _, err := r.buildAndValidateBMCCredentials(request, host)
+	if err != nil {
+		return prov, ready, errors.Wrap(err, "failed to get bmc secret")
+	}
+
```

Comment:
Since this controller only communicates with the Ironic APi via gophercloud and doesn't need to access the BMC directly (for power cycle etc), I'd suggest we create the provisioner without BMC info, similar to what we do for the hostFirmwareSettings controller (https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/hostfirmwaresettings_controller.go#L139). This will simplify creation of the provisioner and will no longer need buildAndValidateBMCCredentials().